compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -109,12 +109,25 @@ object RefChecks {
       for (reqd <- cinfo.cls.givenSelfType.classSymbols)
         checkSelfConforms(reqd, "missing requirement", "required")
 
+      def illegalEnumFlags = !cls.isOneOf(Enum | Trait)
+      def isJavaEnum = parents.exists(_.classSymbol == defn.JavaEnumClass)
+
       // Prevent wrong `extends` of java.lang.Enum
-      if !migrateTo3 &&
-         !cls.isOneOf(Enum | Trait) &&
-         parents.exists(_.classSymbol == defn.JavaEnumClass)
-      then
+      if !migrateTo3 && illegalEnumFlags && isJavaEnum then
         report.error(CannotExtendJavaEnum(cls), cls.sourcePos)
+      else if illegalEnumFlags && isJavaEnum then
```

Comment:
A comment explaining why we need to do something different in the migration code path would be nice

Reply:
done